### ISSUE TITLE
Lower group_norm to XLA

### ISSUE THREAD

**Issue by ailzhang** [pytorch/xla#2155]
As discussed in https://github.com/pytorch/xla/pull/2150

**Comment by harpone** [comment]
I'd be interested in hearing how this affects performance (in resnet50 size nets and above), since I'm using group norm now (which seems to be working fine).

**Comment by ailzhang** [comment]
@harpone #2150 should not impact perf as it generates the exactly same HLO as we did before. 
If we lower group_norm in the future(which will close this issue), I'd expect that implementation to be at least faster than the current one.

**Comment by harpone** [comment]
OK good to know, thanks! Is it possible that not having a lowered group_norm would affect stability? I'm now getting NaNs (after several thousands of iterations) when using group_norm in my resnet instead of batch norm...

**Comment by dlibenzi** [comment]
How did you get a group_norm not lowered?
The pytorch PR did not go in yet AFAIK.

**Comment by dlibenzi** [comment]
If you did not go through the XLA lowering, your execution would be really slow.

So you would really notice it.

**Comment by harpone** [comment]
Yeah I was using a really big resnet and thought the slowness was just that, but now I was testing with batch norm and that's about 4-5 times faster :/ so yeah no group norm for me yet

**Comment by dlibenzi** [comment]
Do you mind printing the metrics after 8..10 steps?

**Comment by harpone** [comment]
Here's the report after 80 iterations. FYI it has also "Weight standardization" conv layers as per the recent WS paper.

```
Metric: CompileTime
  TotalSamples: 4
  Accumulator: 817ms167.064us
  ValueRate: 012ms677.096us / second
  Rate: 0.0571589 / second
  Percentiles: 1%=029ms274.274us; 5%=029ms274.274us; 10%=029ms274.274us; 20%=029ms274.274us; 50%=263ms032.630us; 80%=276ms825.148us; 90%=276ms825.148us; 95%=276ms825.148us; 99%=276ms825.148us
Metric: DeviceLockWait
  TotalSamples: 114
  Accumulator: 12s552ms187.666us
  ValueRate: 062ms010.983us / second
  Rate: 0.611941 / second
  Percentiles: 1%=001.903us; 5%=002.320us; 10%=002.773us; 20%=003.068us; 50%=003.555us; 80%=004.660us; 90%=007.842us; 95%=01s429ms861.468us; 99%=01s449ms983.403us
Metric: ExecuteTime
  TotalSamples: 89
  Accumulator: 02m06s474ms200.170us
  ValueRate: 679ms205.785us / second
  Rate: 0.477958 / second
  Percentiles: 1%=008ms476.705us; 5%=01s423ms853.617us; 10%=01s424ms549.381us; 20%=01s425ms460.583us; 50%=01s430ms869.538us; 80%=01s447ms674.853us; 90%=01s452ms037.788us; 95%=01s455ms980.165us; 99%=02s544ms264.965us
Metric: InboundData
  TotalSamples: 24
  Accumulator: 96.00B
  ValueRate: 0.92B / second
  Rate: 0.230188 / second
  Percentiles: 1%=4.00B; 5%=4.00B; 10%=4.00B; 20%=4.00B; 50%=4.00B; 80%=4.00B; 90%=4.00B; 95%=4.00B; 99%=4.00B
Metric: InputOutputAliasCount
  TotalSamples: 4
  Accumulator: 2667.00
  ValueRate: 38.26 / second
  Rate: 0.0573823 / second
  Percentiles: 1%=204.00; 5%=204.00; 10%=204.00; 20%=204.00; 50%=820.00; 80%=823.00; 90%=823.00; 95%=823.00; 99%=823.00
Metric: IrValueTensorToXlaData
  TotalSamples: 204
  Accumulator: 742ms574.447us
  ValueRate: 910ms803.751us / second
  Rate: 250.278 / second
  Percentiles: 1%=823.447us; 5%=872.472us; 10%=979.590us; 20%=001ms055.507us; 50%=001ms411.209us; 80%=004ms845.979us; 90%=008ms101.023us; 95%=014ms310.396us; 99%=038ms990.723us
Metric: OutboundData
  TotalSamples: 437
  Accumulator: 1.87GB
  ValueRate: 10.22MB / second
  Rate: 2.32688 / second
  Percentiles: 1%=4.00B; 5%=4.00B; 10%=4.00B; 20%=4.00B; 50%=512.00B; 80%=8.00KB; 90%=4.00MB; 95%=73.50MB; 99%=73.50MB
Metric: ReleaseDataHandlesTime
  TotalSamples: 2137
  Accumulator: 04s229ms192.011us
  ValueRate: 031ms485.863us / second
  Rate: 18.5796 / second
  Percentiles: 1%=396.275us; 5%=438.792us; 10%=478.269us; 20%=569.895us; 50%=897.614us; 80%=002ms858.356us; 90%=003ms192.226us; 95%=006ms308.160us; 99%=013ms871.086us
Metric: TensorsGraphSize
  TotalSamples: 90
  Accumulator: 1047391.00
  ValueRate: 5623.64 / second
  Rate: 0.483227 / second
  Percentiles: 1%=1020.00; 5%=11763.00; 10%=11763.00; 20%=11763.00; 50%=11763.00; 80%=11763.00; 90%=11763.00; 95%=11772.00; 99%=11772.00
Metric: TransferFromServerTime
  TotalSamples: 24
  Accumulator: 070ms507.907us
  ValueRate: 666.662us / second
  Rate: 0.230188 / second
  Percentiles: 1%=753.015us; 5%=806.954us; 10%=820.341us; 20%=899.770us; 50%=001ms033.795us; 80%=005ms729.447us; 90%=005ms179.160us; 95%=008ms446.500us; 99%=021ms303.498us
Metric: TransferToServerTime
  TotalSamples: 437
  Accumulator: 02m24s444ms869.206us
  ValueRate: 763ms499.361us / second
  Rate: 2.30989 / second
  Percentiles: 1%=822.327us; 5%=943.536us; 10%=001ms048.441us; 20%=001ms238.135us; 50%=002ms678.427us; 80%=01s354ms438.629us; 90%=01s381ms993.876us; 95%=01s418ms902.976us; 99%=01s482ms349.409us
Metric: TransferToServerTransformTime
  TotalSamples: 437
  Accumulator: 376ms669.978us
  ValueRate: 002ms000.318us / second
  Rate: 2.32688 / second
  Percentiles: 1%=049.992us; 5%=064.658us; 10%=091.885us; 20%=122.499us; 50%=153.207us; 80%=238.044us; 90%=620.759us; 95%=008ms701.135us; 99%=012ms469.582us
Counter: CachedCompile
  Value: 86
Counter: CreateCompileHandles
  Value: 4
Counter: CreateDataHandles
  Value: 72993
Counter: CreateXlaTensor
  Value: 409787
Counter: DestroyDataHandles
  Value: 72234
Counter: DestroyXlaTensor
  Value: 408951
Counter: MarkStep
  Value: 90
Counter: ReleaseDataHandles
  Value: 72235
Counter: UncachedCompile
  Value: 4
Counter: XRTAllocateFromTensor_Empty
  Value: 41
Counter: XrtCompile_Empty
  Value: 144
Counter: XrtExecuteChained_Empty
  Value: 144
Counter: XrtExecute_Empty
  Value: 144
Counter: XrtRead_Empty
  Value: 144
Counter: XrtReleaseAllocationHandle_Empty
  Value: 144
Counter: XrtReleaseCompileHandle_Empty
  Value: 144
Counter: XrtSessionCount
  Value: 11
Counter: XrtSubTuple_Empty
  Value: 144
Counter: aten::_local_scalar_dense
  Value: 24
Counter: xla::_log_softmax
  Value: 90
Counter: xla::_log_softmax_backward_data
  Value: 90
Counter: xla::add
  Value: 16290
Counter: xla::add_
  Value: 54469
Counter: xla::addcdiv_
  Value: 18156
Counter: xla::addcmul
  Value: 4500
Counter: xla::addcmul_
  Value: 18157
Counter: xla::addmm
  Value: 180
Counter: xla::argmax
  Value: 8
Counter: xla::as_strided
  Value: 204
Counter: xla::avg_pool2d
  Value: 90
Counter: xla::avg_pool2d_backward
  Value: 90
Counter: xla::convolution_backward_overrideable
  Value: 4500
Counter: xla::convolution_overrideable
  Value: 4500
Counter: xla::copy_
  Value: 572
Counter: xla::div
  Value: 49665
Counter: xla::empty
  Value: 5570
Counter: xla::empty_strided
  Value: 204
Counter: xla::eq
  Value: 8
Counter: xla::expand
  Value: 18000
Counter: xla::fill_
  Value: 90
Counter: xla::mean
  Value: 18008
Counter: xla::mm
  Value: 360
Counter: xla::mul
  Value: 45360
Counter: xla::mul_
  Value: 36314
Counter: xla::native_batch_norm
  Value: 4500
Counter: xla::native_batch_norm_backward
  Value: 4500
Counter: xla::neg
  Value: 9000
Counter: xla::nll_loss_backward
  Value: 90
Counter: xla::nll_loss_forward
  Value: 90
Counter: xla::relu
  Value: 4500
Counter: xla::select
  Value: 360
Counter: xla::slice
  Value: 360
Counter: xla::sqrt
  Value: 18157
Counter: xla::std
  Value: 4500
Counter: xla::sub
  Value: 9000
Counter: xla::sum
  Value: 18180
Counter: xla::t
  Value: 720
Counter: xla::threshold_backward
  Value: 4500
Counter: xla::unsqueeze
  Value: 4500
Counter: xla::view
  Value: 54180
Counter: xla::zero_
  Value: 18924
Metric: XrtAllocateFromTensor
  TotalSamples: 11538
  Accumulator: 35s073ms686.814us
  Mean: 007ms918.081us
  StdDev: 012ms574.410us
  Rate: 22.1465 / second
  Percentiles: 25%=351.236us; 50%=500.026us; 80%=023ms923.100us; 90%=028ms561.809us; 95%=030ms213.238us; 99%=037ms649.139us
Metric: XrtCompile
  TotalSamples: 80
  Accumulator: 17m44s291ms514.758us
  Mean: 13s554ms631.434us
  StdDev: 19s965ms540.270us
  Rate: 0.122905 / second
  Percentiles: 25%=005ms583.648us; 50%=105ms688.742us; 80%=42s930ms186.078us; 90%=43s752ms837.897us; 95%=43s796ms141.005us; 99%=43s808ms978.304us
Metric: XrtExecute
  TotalSamples: 1088
  Accumulator: 26m34s016ms356.586us
  Mean: 01s440ms207.872us
  StdDev: 233ms999.843us
  Rate: 2.61408 / second
  Percentiles: 25%=01s425ms129.240us; 50%=01s429ms775.550us; 80%=01s446ms688.327us; 90%=01s451ms162.192us; 95%=01s457ms719.903us; 99%=02s549ms025.747us
Metric: XrtExecutorEvict
  TotalSamples: 0
  Accumulator: nanB
  Mean: nanB
  StdDev: nanB
  Percentiles: 
Metric: XrtReadLiteral
  TotalSamples: 288
  Accumulator: 106ms050.354us
  Mean: 368.230us
  StdDev: 239.333us
  Rate: 0.874804 / second
  Percentiles: 25%=268.044us; 50%=311.662us; 80%=411.808us; 90%=513.021us; 95%=591.659us; 99%=002ms109.637us
Metric: XrtReleaseAllocation
  TotalSamples: 23389
  Accumulator: 09s176ms761.782us
  Mean: 541.633us
  StdDev: 629.251us
  Rate: 85.8103 / second
  Percentiles: 25%=052.373us; 50%=260.279us; 80%=001ms076.488us; 90%=002ms525.677us; 95%=002ms908.161us; 99%=002ms405.963us
```

**Comment by dlibenzi** [comment]
Looks good. Group norm is routed to batch-norm, and there are no recompilations.
There are 24 `aten::_local_scalar_dense` in 90 step (`MarkStep`) which are likely `item()` calls.
How many steps between loss reporting?
Step time is about 1.4s, which model is this?

**Comment by harpone** [comment]
I was doing every 10 iterations, which I usually do when debugging (every 100 iters in real cases).

Uhh this is a resnet variant, but can't say about number of params at the moment... somewhat bigger than resnet50.

**Comment by dlibenzi** [comment]
Every 10+ iterations should be fine.
If you end up collecting many metrics to report (many `item()` calls), I suggest moving your reporting in a step closure:

https://github.com/pytorch/xla/blob/e6c163d3bd39de072041d2ef6954453db41b23f4/test/test_train_mp_mnist.py#L135

Remember to pass the tensors as args, and not calling `item()` at the call site.
